src/nlr/DeepPolyMaxPoolElement.cpp

```diff
@@ -0,0 +1,227 @@
+/*********************                                                        */
+/*! \file DeepPolyMaxPoolElement.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Haoze Andrew Wu
+ ** This file is part of the Marabou project.
+ ** Copyright (c) 2017-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved. See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** [[ Add lengthier description here ]]
+
+**/
+
+#include "DeepPolyMaxPoolElement.h"
+#include "FloatUtils.h"
+
+namespace NLR {
+
+DeepPolyMaxPoolElement::DeepPolyMaxPoolElement( Layer *layer )
+{
+    _layer = layer;
+    _size = layer->getSize();
+    _layerIndex = layer->getLayerIndex();
+}
+
+DeepPolyMaxPoolElement::~DeepPolyMaxPoolElement()
+{
+    freeMemoryIfNeeded();
+}
+
+void DeepPolyMaxPoolElement::execute( const Map<unsigned, DeepPolyElement *>
+                               &deepPolyElementsBefore )
+{
+    log( "Executing..." );
+    ASSERT( hasPredecessor() );
+    allocateMemory();
+
+    // Update the symbolic and concrete upper- and lower- bounds
+    // of each neuron
+    for ( unsigned i = 0; i < _size; ++i )
+    {
+        log( Stringf( "Handling Neuron %u_%u...", _layerIndex, i ) );
+        List<NeuronIndex> sources = _layer->getActivationSources( i );
+
+        NeuronIndex indexOfMaxLowerBound = *( sources.begin() );
+        double maxLowerBound = FloatUtils::negativeInfinity();
+        double maxUpperBound = FloatUtils::negativeInfinity();
+
+        Map<NeuronIndex, double> sourceLbs;
+        Map<NeuronIndex, double> sourceUbs;
+        for ( const auto &sourceIndex : sources )
+        {
+            DeepPolyElement *predecessor =
+                deepPolyElementsBefore[sourceIndex._layer];
+            double sourceLb = predecessor->getLowerBound
+                ( sourceIndex._neuron );
+            sourceLbs[sourceIndex] = sourceLb;
+            double sourceUb = predecessor->getUpperBound
+                ( sourceIndex._neuron );
+            sourceUbs[sourceIndex] = sourceUb;
+
+            if ( maxLowerBound < sourceLb )
+            {
+                indexOfMaxLowerBound = sourceIndex;
+                maxLowerBound = sourceLb;
+            }
+            if ( maxUpperBound < sourceUb )
+            {
+                maxUpperBound = sourceUb;
+            }
+        }
+
+        // The phase is fixed if the lower-bound of a source variable x_b is
+        // larger than the upper-bounds of the other source variables.
+        bool phaseFixed = true;
+        for ( const auto &sourceIndex : sources )
+        {
+            if ( sourceIndex != indexOfMaxLowerBound &&
+                 FloatUtils::gt( sourceUbs[sourceIndex], maxLowerBound ) )
```

Comment:
shouldn't this be lt?

Reply:
Whoops yes. I'll add more test for max constraints.

Reply:
> Whoops yes. I'll add more test for max constraints.

Actually. I think the code is right. The max constraint is fixed if there is a input variable whose lower bound is greater than the upper bound of the other input variables. In that case, we can conclude that the Max is fixed.